Documentation/design-docs/AssemblyLoadContext.Activate.md

```diff
@@ -0,0 +1,306 @@
+# AssemblyLoadContext.Activate
+
+The `AssemblLoadContext` is designed to create a assembly load scope and type resolving scope.
+
+It is intended as a replacement for the isolation offered by `AppDomain`s.
+
+Current APIs affect and/or use a given `AssemblyLoadContext` by:
+* Operating directly on the `AssemblyLoadContext`
+* Operating on an assembly loaded into the `AssemblyLoadContext`
+* Executing methods defined within an assembly loaded by the `AssemblyLoadContext`
+
+There is a need for a general solution to allow code executing outside a given `AssemblyLoadConxt` to behave as if it was loaded within the context.
+
+This is in part because of the design of `AssemblyLoadContext` which
+* Requires `System.Private.CoreLib` to be loaded exclusively in the `Default` context.
+* Strongly recommends the runtime framework be completely loaded into the `Default` context.
+* Recommends most frameworks be loaded exclusively in the `Default` context.
+
+This proposal is for a mechanism for code to explicitly set a specific `AssemblyLoadContext` as the active context.
+
+## Proposed API extension(s)
+```C#
+namespace System.Runtime
+{
+  partial class AssemblyLoadContext
```

Comment:
This is very similar to the [Activation Context](https://docs.microsoft.com/windows/desktop/sbscs/using-the-activation-context-api) in COM.

Reply:
> This is very similar to the Activation Context in COM.

@AaronRobinsonMSFT 
Is that good or bad?

In what way is `AssemblyLoadContext` similar to [Activation Context](https://docs.microsoft.com/windows/desktop/sbscs/using-the-activation-context-api)

I am not even a COM beginner, and the [Activation Context](https://docs.microsoft.com/windows/desktop/sbscs/using-the-activation-context-api) doc is not very enlightening.

Reply:
It is good in the sense we can argue there is value because COM did it and it is a proven model and a scenario exists for the concept. It is bad because there it demonstrates a discontinuity in the ALC model if it exists.

In COM, a context is _not_ associated with an assembly/library/binary. The activation context is bound to the thread. Most if not all COM security/activation/management is entirely based on thread affinity. There are scenarios where this is very useful (i.e. implicit MTA), but there are also cases where it makes the world a much more difficult place (i.e. the intersection between STA and RCW/GC). We should really talk since this is a very complicated area and I don't want to type a treatise on all the ramifications.

You can think of a COM Activation Context as exactly what you are proposing here, except it is bound to the thread (stored in TLS). There is a stack of contexts that have an implied hierarchy/scope that are tried in order. The context stores a list of assemblies (not .NET assemblies, but rather SxS) that are interrogated for specific class IDs for COM activation. If the collection of activation contexts associated with the thread don't have a record for the class ID in question, then the COM system uses the global registration (i.e. Registry). This is also probably something we should chat about.

Reply:
I do not think that this is same as COM activation context.

This API is designed to specifically address problem with some of the existing reflection APIs that infer context via stackwalk from immediate caller. Maybe the API should have "Reflection" in the name to make it clear, e.g. `ContextForReflection`.

COM never infers context via stackwalk from the caller assembly. The problem solved by COM activation context is different.

Reply:
@jkotas In this case we are scoping the behavior but the COM Activation Context is basically the general approach that this implies i.e. current thread and can be controlled via push/pop stack semantics - the example is what I did in [CoreRun](https://github.com/dotnet/coreclr/blob/master/src/coreclr/hosts/corerun/corerun.cpp#L639-L650). They are functionally equivalent and imply a type of thread affinity. The COM version has broader functional abilities, but the idea and intent is the same - scoped executional context.